examples/supplementary/arg_iter/src/lib.rs

```diff
@@ -0,0 +1,192 @@
+        let generics = cx.tcx.generics_of(id);
+
+        // No generics on this function
+        if generics.own_params.is_empty() {
+            return;
+        }
+
+        let predicates = cx.tcx.predicates_of(id);
+
+        // Find generic type parameters that are *only* bound by Iterator (+ implicit Sized)
+        let mut iterator_only_params: Vec<Symbol> = Vec::new(); // Store only the name (Symbol)
+        for param in &generics.own_params {
+            // Iterate over a slice
+            if let ty::GenericParamDefKind::Type { .. } = param.kind {
+                // param.def_id is already the DefId
+                if is_param_bound_only_by_iterator(cx, predicates, param.def_id, iterator_def_id) {
+                    iterator_only_params.push(param.name); // Push only the name
+                }
+            }
+        }
+
+        if iterator_only_params.is_empty() {
+            return; // No relevant generic parameters found
+        }
+
+        // Check function inputs using rustc_middle types for robustness
+        let fn_sig = cx.tcx.fn_sig(id).skip_binder(); // Use skip_binder to handle late-bound lifetimes
+
+        for (i, input_ty) in fn_sig.inputs().iter().enumerate() {
+            // Check if the input type is a generic parameter (TyKind::Param)
+            if let ty::TyKind::Param(param_ty) = input_ty.skip_binder().kind() {
+                // Map the ParamTy index back to the GenericParamDef
+                // Use param_at since opt_param_at is deprecated and the index should be valid
+                let param_def = generics.param_at(param_ty.index as usize, cx.tcx);
+
+                // Check if this parameter's name is in our list of iterator-only params
+                if iterator_only_params.contains(&param_def.name) {
+                    // Get the span from the corresponding HIR FnDecl input type for accurate
+                    // reporting
+                    let input_hir_ty = &decl.inputs[i];
+                    span_lint_and_help(
+                        cx,
+                        ARG_ITER,
+                        input_hir_ty.span, /* Lint the type in the function signature (using
+                                            * HIR span) */
+                        "parameter type has `Iterator` bound",
+                        None,
+                        format!(
+                            "consider using `IntoIterator` instead of `Iterator` for parameter `{}`",
+                            param_def.name
+                        ),
+                    );
+                }
+            }
+        }
+    }
+}
+
+/// Checks if a given type parameter `param_def_id` is bound _only_ by `iterator_def_id`
+/// within the given `predicates`. Implicit `Sized` bounds are ignored.
+fn is_param_bound_only_by_iterator<'tcx>(
+    cx: &LateContext<'tcx>,
+    predicates: ty::GenericPredicates<'tcx>,
+    param_def_id: DefId, // The DefId of the type parameter we are checking (e.g., `I`)
```

Comment:
I think it should be possible to pass a `rustc_middle::ty::Ty` here, rather than a `DefId`.